optimizer: support early finalization

aviatesk opened the pull request:
Currently, in the finalizer inlining pass, if not all the code between the finalizer registration and the end of the object’s lifetime (i.e., where the finalizer would be inlined) is marked as `:nothrow`, it simply bails out. However, even in such cases, we can insert a `finalize` call at the end of the object’s lifetime, allowing us to call the finalizer early if no exceptions occur.

This commit implements this optimization. To do so, it also moves `finalize` to `Core`, so the compiler can handle it directly.

aviatesk committed:
optimizer: early `finalize` insertion

Currently, in the finalizer inlining pass, if not all the code between
the finalizer registration and the end of the object’s lifetime (i.e.,
where the finalizer would be inlined) is marked as `:nothrow`, it simply
bails out. However, even in such cases, we can insert a `finalize` call
at the end of the object’s lifetime, allowing us to call the finalizer
early if no exceptions occur.

This commit implements this optimization. To do so, it also moves
`finalize` to `Core`, so the compiler can handle it directly.

vtjnash commented on base/compiler/ssair/passes.jl:
```diff
@@ -1648,18 +1647,23 @@ function try_resolve_finalizer!(ir::IRCode, alloc_idx::Int, finalizer_idx::Int,
+            # where the object’s lifetime ends (`insert_idx`): In such cases, we can’t
+            # remove the finalizer registration, but we can still insert a `Core.finalize`
+            # call at `insert_idx` while leaving the registration intact.
+            newinst = add_flag(NewInstruction(Expr(:call, GlobalRef(Core, :finalize), finalizer_stmt.args[3]), Nothing), flag)
```
This can have a surprising performance cost, since running finalizers are normally O(1), but this call is O(n). We would need to reimplement the way finalizers are registered.

To make this particular case legal, we also need to prove that we are only going to run the expected finalizer that it originally detected being registered (no other finalizers can be registered along side it, though those may be considered lifetime escapes already?) as other finalizers might not be allowed to run on this Task. There is also an issue that inference did the analysis in the tls world, but this runs them in the unknown future world.

aviatesk replied:
> To make this particular case legal, we also need to prove that we are only going to run the expected finalizer that it originally detected being registered (no other finalizers can be registered along side it, though those may be considered lifetime escapes already?

Regarding this, the escape check in `sroa_mutables!` ensures that the object in question only has a single finalizer.

> There is also an issue that inference did the analysis in the tls world, but this runs them in the unknown future world.

The docstring for `finalizer` states:

> > Note that there is no guaranteed world age for the execution of `f`. It may be called in the world age in which the finalizer was registered or any later world age.

So, it should be acceptable for the world in which the finalizer is inferred and the world in which the finalizer is executed to be different?

vtjnash replied:
> Regarding this, the escape check in sroa_mutables! ensures that the object in question only has a single finalizer.

Okay good. I wasn't sure if the analysis was total (guaranteed that there was no other finalizer), or simply that it only uses one of the finalizers, depending on what it can detect locally

> So, it should be acceptable for the world in which the finalizer is inferred and the world in which the finalizer is executed to be different?

It is acceptable to be different, but therefore not acceptable for inference to make decisions based on its guaranteed behavior, and then to not enforce the world of the runtime. We could probably change `finalize`, but the main concern is that `finalize` is already slow and undesirable compared to letting the GC do its work to sweep them as a batch, and we don't necessarily want to make `finalizer` slower just to support that function. So we can try to forward the function in `finalizer` directly here, but we need some (fast) way to also undo that finalizer registration (or delay it until either an error happens or some other branch which won't execute this finalize call insertion location)

aviatesk replied:
Yeah, it looks like `finalize` is pretty slow. I tried to optimize it to some extent in 754a8bdd613cad562312e222a121c43f542bcec3, but it’s still about 2.5 times slower compared to `master` in the following benchmark:
```julia
​mutable struct AtomicCounter
    @atomic count::Int
end

const counter = AtomicCounter(0)

function withfinalizer(x)
    xs = finalizer(Ref(x)) do obj
        Base.@assume_effects :nothrow :notaskstate
        @atomic counter.count += obj[]
    end
    println(devnull, xs[])
    return xs[]
end

@benchmark withfinalizer(1)
```
(even though I confirmed that `withfinalizer(1)` certainly hits the fast path added in the commit)

So, as you mentioned, we’ll probably need to implement a way to cancel the finalizer registration or delay the registration itself. The former sounds a bit simpler, so I’ll explore that approach.

aviatesk replied:
I tried out the approach of canceling the finalizer registration (0f1da0e3d7f567a8b0f18c517cd7a0de87c8e61c). What do you think?

aviatesk committed:
optimizer: cancel finalizer registration rather than inserting `finalize`

This seems to perform well.

```julia
mutable struct AtomicCounter
    @atomic count::Int
end
const counter = AtomicCounter(0)

function withfinalizer(x)
    xs = finalizer(Ref(x)) do obj
        Base.@assume_effects :nothrow :notaskstate
        @atomic counter.count += obj[]
    end
    println(devnull, xs[])
    xs[] = xs[]
    println(devnull, xs[])
    return xs[]
end

@benchmark withfinalizer(100)
```

> master
```
BenchmarkTools.Trial: 10000 samples with 867 evaluations.
 Range (min … max):  135.717 ns …  18.617 μs  ┊ GC (min … max):  0.00% … 47.51%
 Time  (median):     144.992 ns               ┊ GC (median):     0.00%
 Time  (mean ± σ):   192.949 ns ± 687.513 ns  ┊ GC (mean ± σ):  11.57% ±  3.23%

     ▇  ▃█▅▂
  ▁▁▂█▇▂████▇▆▅▄▄▄▅▄▄▂▂▂▂▂▁▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▂
  136 ns           Histogram: frequency by time          201 ns <

 Memory estimate: 208 bytes, allocs estimate: 7.
```

> this commit
```
BenchmarkTools.Trial: 10000 samples with 964 evaluations.
 Range (min … max):   83.506 ns …   3.408 μs  ┊ GC (min … max):  0.00% … 96.85%
 Time  (median):      88.780 ns               ┊ GC (median):     0.00%
 Time  (mean ± σ):   101.797 ns ± 137.067 ns  ┊ GC (mean ± σ):  10.26% ±  7.31%

    ▄█▃▅
  ▄▃████▅▅▆▅▄▄▄▃▃▃▃▃▃▃▃▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▁▂▁▂▂▁▁▁▁▁▁▂▁▂▂▁▂▂ ▃
  83.5 ns          Histogram: frequency by time          145 ns <

 Memory estimate: 208 bytes, allocs estimate: 7.
```

aviatesk commented:
Instead of simply inserting `finalize`, I changed the approach to inline the finalizer body and cancel the finalizer registration (by preparing new builtin `Core._cancel_finalizer`).
It looks like this PR is now achieving pretty good performance.
```julia
mutable struct AtomicCounter
    @atomic count::Int
end
const counter = AtomicCounter(0)
const _throw_or_not = Ref(false)
@noinline throw_or_noop() = _throw_or_not[] ? error("") : nothing
function withfinalizer(x)
    xs = finalizer(Ref(x)) do obj
        Base.@assume_effects :nothrow :notaskstate
        @atomic counter.count += obj[]
    end
    throw_or_noop()
    return xs[]+=1
end
@benchmark withfinalizer(0)
```

> master
```
julia> @benchmark withfinalizer(1)
BenchmarkTools.Trial: 10000 samples with 999 evaluations.
 Range (min … max):  11.177 ns … 124.200 μs  ┊ GC (min … max):  0.00% … 39.95%
 Time  (median):     11.928 ns               ┊ GC (median):     0.00%
 Time  (mean ± σ):   46.081 ns ±   1.765 μs  ┊ GC (mean ± σ):  30.60% ±  0.84%

   ▃▄▅▇▇█▆▆▄▅▆▆▅▅▃▃▂▁ ▁                 ▁▁ ▁▂▃▁▁▁ ▁            ▃
  ▇██████████████████████▇█▇█▇█▇▇▆▇▇█▇█▇███████████▇▆▇▆▄▆▃▄▁▄▄ █
  11.2 ns       Histogram: log(frequency) by time      17.4 ns <

 Memory estimate: 16 bytes, allocs estimate: 1.
```

> this PR
```
julia> @benchmark withfinalizer(1)
BenchmarkTools.Trial: 10000 samples with 998 evaluations.
 Range (min … max):  17.327 ns …  1.416 μs  ┊ GC (min … max): 0.00% … 97.78%
 Time  (median):     18.329 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   18.802 ns ± 16.372 ns  ┊ GC (mean ± σ):  1.35% ±  1.65%

           ▂█▇▂                                                
  ▂▂▂▂▂▂▃▅▆████▇▅▃▂▂▃▄▄▄▃▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▁▂▂▁▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂ ▃
  17.3 ns         Histogram: frequency by time        22.6 ns <

 Memory estimate: 16 bytes, allocs estimate: 1.
```

vtjnash commented on src/gc-common.c:
```diff
@@ -478,6 +480,45 @@ JL_DLLEXPORT void jl_finalize(jl_value_t *o)
+}
+
+// Remove the finalizer for `o` from `ct->ptls->finalizers` and cancel that finalizer.
+// `lookup_idx` is the index at which the finalizer was registered in `ct->ptls->finalizers`
```
This will just be a random pointer though. After the next GC we will have no idea where the finalizer index will be without scanning for it.

aviatesk replied:
Yes. If a GC occurs between the finalizer registration and the call to `Core._cancel_finalizer`, this index becomes meaningless, but in that case, the slow path of `jl_cancel_finalizer` will handle it by performing a full scan. `lookup_idx` is only used for the fast path.

vtjnash replied:
But doesn't that means we expect this to normally hit the slow case, except when microbenchmarking?

oscardssmith replied:
no. There are plenty of benchmarks (e.g. BigInt) where we expect many objects to die before GC has run.

vtjnash reviewed (commented):
I am still very unsure that we should do this. It could completely change the order in which finalizers are run, since currently they can be assumed to run in LIFO dominator order (e.g. DFS post-order), but afterwards, we may lose that assumed ordering?

aviatesk commented:
This PR implements an optimization that inlines the target finalizer call and, instead of removing the finalizer registration like before, cancels the registration by calling `Core._cancel_finalizer`. With finalizer cancellation, only the target finalizer (and its corresponding object) is removed from the finalizers list, so I don't think it changes the order of other finalizers. Am I misunderstanding something here?

aviatesk committed:
call `_cancel_finalizer` a bit earlier

vtjnash commented on src/gc-common.c:
```diff
@@ -478,6 +477,46 @@ JL_DLLEXPORT void jl_finalize(jl_value_t *o)
+    void **items = list->items;
+    void *v = items[idx];
+    if (o == (jl_value_t*)gc_ptr_clear_tag(v, 1)) {
+        for (size_t j = idx + 2; j < list->len; j += 2) {
+            items[j-2] = items[j];
+            items[j-1] = items[j+1];
+        }
+        list->len = list->len - 2;
```
This also invalidates all the other `lookup_idx` values, and is potentially `O(n)` slow. We should at least only NULL the fptr and discard it on the next sweep

aviatesk replied:
Implemented.

aviatesk committed:
ignore `ThrownEscape`

aviatesk committed:
just null out cancelled finalizers

vtjnash reviewed (commented):
Comments from offline discussion:
 - we could allow this for any finalizer function, not just ones with favorable effects
 - we should usually use the shim in gc-common.c to call the finalizers, so that they also may respect the finalizers-disabled flag. This could even be integrated with the call to cancel it
 - when canceling a finalizer, we can use the idx as a hint for the index to find it at and scan backwards from there in the list until we find the right one to cancel
 - after cancelling a finalizer, we can trim any trailing null values from the ptls list, to avoid the index from getting too far off in some typical cases
 - even with this, we are still guaranteeing that inner objects with finalizers registered before outer objects with finalizers will run in LIFO order, since `finalizer` itself is an escape and must be treated as such (even if it gets removed by optimizations later) and so inner objects cannot die before their containers
